### ISSUE TITLE
Scopes defined in  appsettings.json being wrongly validated?

### ISSUE THREAD

**Issue by pealmeid** [AzureAD/microsoft-identity-web#1600]
Hello,

I'm trying to move my scopes declarations from code to `appsettings.json` but I'm having trouble making it work. I noticed a strange behavior and hope to understand if I'm doing something wrong or if this is an actual issue.

The methods in my API controller were being validated as expected when annotated with:
```csharp
RequiredScope("scope_name")
public class ApiController : Controller
{
    public async Task<IActionResult> Method()
    {
    }
}
```

But I changed them to get the scope names from a key in `appsettings.json`:
```csharp
RequiredScope(RequiredScopesConfigurationKey = "AzureAd:Scopes")
public class ApiController : Controller
{
    public async Task<IActionResult> Method()
    {
    }
}
```

And added the corresponding key to `appsettings.json`:
```json
{
    "AzureAd": {
        "Instance": "https://login.microsoftonline.com/",
        "TenantId": "00000000-0000-0000-0000-000000000000",
        "ClientId": "00000000-0000-0000-0000-000000000000",
        "Scopes": "scope_name"
    }
}
```

However, after I made this change I noticed that `ApiController.Method` is validated no matter what value I put in the `AzureAd:Scopes` key, or even if I don't put the key in `appsettings.json` at all.

To summarize:

| Method Annotation | Scope in `appsettings.json` | Method Validated? | As Expected? |
|---|:---:|:---:|:---: |
| `RequiredScope("right_scope_name")` | - | YES | YES |
| `RequiredScope("wrong_scope_name")` | - | NO | YES |
| `RequiredScope(RequiredScopesConfigurationKey = "AzureAd:Scopes")` | `right_scope_name` | YES | YES |
| `RequiredScope(RequiredScopesConfigurationKey = "AzureAd:Scopes")` | `wrong_scope_name` | YES | **NO** |
| `RequiredScope(RequiredScopesConfigurationKey = "AzureAd:Scopes")` | empty | YES | **NO** |

I cloned the repo to trace through the code and I (sort of) understand what's happening but I'm not confident enough to say it's a bug instead of some mistake on my part.

Any help is greatly appreciated, and please feel free to ask for any additional information.

**Comment by jmprieur** [comment]
@pealmeid :
- which version of Microsoft.Identity.Web are you using?
- do you have repro code (to see the Startup.cs)
Thanks!

**Comment by pealmeid** [comment]
@jmprieur thanks for the quick reply. 

I'm using version  1.22.1  of Microsoft.Identity.Web.

I created a [repo ](https://github.com/pealmeid/msidweb-repro/tree/main) to illustrate the case with an app based on the Visual Studio app template for `ASP.NET Core Web Api.` I did minimal changes to add CORS support, register the authentication service and annotate the methods with the `RequiredScope` attribute.

Of course, to actually run it you still need to create an Azure AD app registration for the API server, as well as a client app to  authenticate the user against the AD and request the token for the scope, but I hope this is enough to illustrate the example.

**Comment by pealmeid** [comment]
I did some research and would like to validate my understanding of required scopes. Please let me know if the following is correct, and apologies in advance if this isn't clear. Any feedback is greatly appreciated.

# Registering the Scopes

There are two ways to register the required scopes for the API: either through a _**policy-based authorization**_ or _**adding metadata**_ to the endpoint.

[1] Policy-based authorization
----
To register the required scopes via policy, you define the policy with `AddAuthorization` in `Program.cs` during setup:
```csharp
builder.Services.AddAuthorization(o =>
{
    o.AddPolicy("RightScopePolicy", policyBuilder => policyBuilder.RequireScope("right_scope"));
});
```
And you apply the policy in one of two ways, either declaratively or through an annotation. The declarative way is through `RequireAuthorization` when you add the endpoint to the controller:
```csharp
app.MapControllers().RequireAuthorization("RightScopePolicy");
```
The other way is through an annotation on the method or controller:
```csharp
[Authorize(Policy = "RightScopePolicy")]
[HttpGet("MethodName")]
public IActionResult Get()
{
    return Ok();
}
```
[2] Endpoint metadata
----
The other way to register the required scopes is adding the requirements as a metadata to the endpoint. This also can be done either declaratively or through annotations. The declarative way is through `RequireScope` when you add the endpoint to the controller:
```csharp
app.MapControllers().RequireScope("right_scope");
```
You can also add the requirements to the metadata through an annotation on the method. This annotation can have two formats:
```csharp
[RequiredScope("right_scope")]
[HttpGet("MethodName")]
public IActionResult Get()
{
    return Ok();
}
```
or
```csharp
[RequiredScope(RequiredScopesConfigurationKey = "AzureAd:Scopes")]
[HttpGet("MethodName")]
public IActionResult Get()
{
    return Ok();
}
```
**Note:** this second annotation format is the one related to this issue. `RequiredScopesConfigurationKey` refers to the section in `appsettings.json` where the scope is defined.

# Handling the Scopes
Once the required scopes are registered, they are handled in the `HandleRequirementAsync` method of `ScopeAuthorizationHandler` (pseudocode below):
```csharp
internal class ScopeAuthorizationHandler : AuthorizationHandler<ScopeAuthorizationRequirement>
{
    //
    protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, ScopeAuthorizationRequirement requirement)
    {
     //
    if (requirements are met) {
        context.Succeed(requirement);
    }

    return Task.CompletedTask;
    }
}
```
`HandleRequirementAsync` receives two parameters, `context` (type `AuthorizationHandlerContext`) and `requirement` (type `ScopeAuthorizationRequirement`).

This is somewhat misleading, because both parameters can contain requirements for the handler, depending on how the scopes are registered (via policy or endpoint).

# My Understanding
From what I could understand:

1. if you use policy-based authorization the required scopes are passed to `HandleRequirementAsync` through the `requirement` parameter;

2. if you use endpoint metadata, the required scopes are passed to `HandleRequirementAsync` through the `context` parameter;

3. the only way to set `RequiredScopesConfigurationKey` is through endpoint metadata, using the `RequiredScope` annotation (it cannot be set declaratively);

4. From [3], `RequiredScopesConfigurationKey` is necessarily passed via the `context` parameter, and extracted on the call below (excerpt from `HandleRequirementAsync`):
```csharp
var endpoint = context.Resource switch
{
    HttpContext httpContext => httpContext.GetEndpoint(),
    Endpoint ep => ep,
    _ => null,
};

var data = endpoint?.Metadata.GetMetadata<IAuthRequiredScopeMetadata>();
```
5. At this point, `data.RequiredScopesConfigurationKey` is valid and contains the key to be used in `appsettings.json` to get the scope;

6. However, `HandleRequirementAsync` does **NOT** check `data.RequiredScopesConfigurationKey` to get the `appsettings.json` key. Instead, it only checks `requirements.RequiredScopesConfigurationKey`. See below an excerpt from `HandleRequirementAsync`:

```csharp
IEnumerable<string>? scopes = null;
if (requirement.RequiredScopesConfigurationKey != null)
{
    scopes = _configuration.GetValue<string>(requirement.RequiredScopesConfigurationKey)?.Split(' ');
}

if (scopes is null)
{
    scopes = requirement.AllowedValues ?? data?.AcceptedScope;
}

// Can't determine what to do without scope metadata, so proceed
if (scopes is null)
{
    context.Succeed(requirement);
    return Task.CompletedTask;
}
```
7. Because (from [4]) `RequiredScopesConfigurationKey` is necessarily passed through the `context` parameter, `requirement.RequiredScopesConfigurationKey` is always `null` and the configuration is not loaded from `appsettings.json`;

8. Also, because `scopes` is `null` at the end of the handler, it assumes that no scope metadata is present and proceeds with `context.Succeed(requirement)`.

# Suggestion
The code in `HandleRequirementAsync` should also check `data.RequiredScopesConfigurationKey` to get the `appsettings,json` configuration:

```csharp
IEnumerable<string>? scopes = null;

var configurationKey = requirement.RequiredScopesConfigurationKey ?? data.RequiredScopesConfigurationKey;
if (configurationKey != null)
{
    scopes = _configuration.GetValue<string>(configurationKey)?.Split(' ');
}

if (scopes is null)
{
    scopes = requirement.AllowedValues ?? data?.AcceptedScope;
}

// Can't determine what to do without scope metadata, so proceed
if (scopes is null)
{
    context.Succeed(requirement);
    return Task.CompletedTask;
}
```
As mentioned before, any feedback is greatly appreciated. Thanks in advance.

**Comment by jmprieur** [comment]
Thanks @pealmeid for reporting this issue and your brilliant analysis. You are perfectly right.
We are going to fix this bug in the next release of Microsoft.identity.Web

**Comment by jennyf19** [comment]
Fixed in [1.22.2 release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.22.2).
